StepContainerV2 DesignPicker: Fix top and bottom bar

zaguiini opened the pull request:
Closes https://github.com/Automattic/wp-calypso/issues/102301.

## Proposed Changes

* Splits the `UnifiedDesignPickerPreview` component out of the giant `UnifiedDesignPickerStep` component
* Removes `shouldHideActionButtons` from the unified design picker component
  * The action buttons in the floating toolbars should always be shown
* Ensures that the sticky bottom bar navigates using the design picker's internal navigation system

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Issue with mobile navigation through the design picker discovered during CfT
pdtkmj-3xe-p2#comment-6772

In the theme preview step on themes without global styles, there should be a navigation header at the top. This PR ensures there's the standard `<TopBar>` in that situation.

But more than that, the sticky bottom bar wasn't plugged in to the design picker's internal navigation system. Which navigates between pages of the design picker without adjusting the URL (so it's disconnected from the usual stepper related navigation).

This resulted in an unfortunately large refactor, and the design picker's navigation mechanism needing to be exposed through a ref https://github.com/Automattic/wp-calypso/pull/102310/files#diff-54190ddf0beadd08aacea4cacaab2a12b490a5e38df6fda709fc4d9fec213f67R13 so the sticky footer could access the navigation functions.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


https://github.com/user-attachments/assets/84985150-c645-4916-8975-ba6c4e7d4825


https://github.com/user-attachments/assets/d86fc807-d674-43b8-9b11-8267ab42d412


* This can be tested on `calypso.live` ([see the link below](https://github.com/Automattic/wp-calypso/pull/102310#issuecomment-2770933343)) but because of the transition between the `onboarding` flow and `site-setup` flow you have to do a little fiddling to disable the `onboarding/step-container-v2` flag (which is what users currently see in production). 
* The StepContainerV2 changes are enabled by default on `calypso.live`.
* If you want to test _without_ StepContainerV2, then you can go through the onboarding flow as usual, then **when you land on the design picker step**, append `&flags=-onboarding/setup-container-v2` to the URL and reload the page.


There are a number of flows to check. These flows should be tested with:
- mobile layout and the `onboarding/step-container-v2` flag set (the default in local dev)
- desktop layout and `onboarding/step-container-v2` set
- mobile layout and `onboarding/step-container-v2` **unset**
- desktop layout and `onboarding/step-container-v2` **unset**

Navigation should work similarly in all these situations.

All flows to check start at `/setup/site-setup/design-setup?siteSlug=%s`. The flows to check are:

| On plans step | On design picker | In design preview | Mobile ✅ w/ flag | Desktop ✅ w/ flag | Mobile ✅ w/o flag | Desktop ✅ w/o flag |
| --- | --- | --- | --- | --- | --- | --- |
| Free | Free theme | Free style variation | ✅ | ✅ | ✅ | ✅ |
| Free | Twenty Twenty-Five | Premium style variation |  ✅ | ✅ |  | ✅ |
| Free | Memphoria | Any style variation | ✅ |  ✅ | ✅ | ✅ |
| Free | Dropp | Make no style changes | ✅ |  ✅ |  | ✅ |
| Free | Dropp | Change font or color, use the save button | ✅ | ✅  |  | ✅ |
| Free | Dropp | Change font or color, use the back button | ✅ |  N/A | ✅ | N/A |
| Free | Chrono (under store designs) | Unlock | ✅ | ✅  | ✅ | ✅ |
| Premium | Free theme | Free style variation | ✅ | ✅  |  | ✅ |
| Premium | Twenty Twenty-Five | Premium style variation | ✅ | ✅   |  | ✅ |
| Premium | Memphoria | Any style variation | ✅ |  ✅ | ✅ | ✅ |
| Premium | Dropp | Make no style changes | ✅ | ✅  |  | ✅ |
| Premium | Dropp | Change font or color, use the save button | ✅ | ✅  |  | ✅ |
| Premium | Dropp | Change font or color, use the back button | ✅ | N/A |  | N/A |
| Business | Chrono (under store designs) | Unlock |   ✅ | ✅  | ✅ | ✅ |



In all of these flows, try using Back to navigate around the design picker/previewer states

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

zaguiini committed:
Add top bar to the theme preview index screen

zaguiini committed:
Use StepContainerV2's bottom bar and buttons within theme preview

zaguiini committed:
Wire bottom bar buttons by passing the navigatorRef

zaguiini committed:
Do not show back button in the bottom bar when browsing design preview index

zaguiini committed:
Split picker and preview into separate components

zaguiini committed:
Do not AsyncLoad preview

zaguiini committed:
Display unlock theme label if user cannot pick the design without paying

zaguiini committed:
Add note about navigatorRef

p-jackson committed:
Rename recordTracksEvent prop, and tracks are recorded by default

p-jackson committed:
Rearrange some hooks in the hopes of making the diff slightly smaller

p-jackson committed:
Another diff tidy up

autumnfjeld commented:
Um, the testing is 48 different tests!   Would it be faster to write a test suite?!?!?  Oh my.

autumnfjeld commented:
* Premium | Chrono (under store designs) | Unlock |  
* Mobile (dev tools responsive)

Tested in Calypso live ( flag on by default)

https://github.com/user-attachments/assets/69b9656c-730f-4e51-b76b-1ba161495cd7


Tested in Calypso live (with flag off :  `&flags=-onboarding/setup-container-v2`)
(I stopped at checkout, since the DP skip button tests were complete)

https://screen.studio/share/WmZrT382


In both these tests I selected a variety of different themes, played wth fonts & colors, going back and forth.  No issues found. ✅

autumnfjeld commented:
✅ 
* Premium | Chrono (under store designs) | Unlock |  
* Desktop


https://www.awesomescreenshot.com/video/38291792?key=f487d0bf81142cf9d21f7a79371798d7

p-jackson committed:
Stop click events from being passed to pickDesign

p-jackson commented:
There was a WSOD when going through the basic flow. We were trying to serialise a click event to local storage and throwing an exception.

https://github.com/user-attachments/assets/6a590609-09b3-4bff-b38d-38b840a3c94d

Fixed here: fa75884b47ea0eb849008fa8469408bafd0c0f2a

TypeScript is very permissive when it comes to checking function types 😢

p-jackson commented:
The primary button in the design picker is incorrect when:
I already have a plan, and I choose Dropp

The button text says "Unlock", but I don't need to unlock anything.

This PR with flag enabled:

![CleanShot 2025-04-02 at 21 51 40@2x](https://github.com/user-attachments/assets/1afbedeb-6707-4936-8362-bf8e3140766b)

Production with flag disabled:

![CleanShot 2025-04-02 at 21 53 13@2x](https://github.com/user-attachments/assets/5dfc4632-f7cc-4cd3-b773-5987bf5c56f2)

p-jackson committed:
Fix button label in design preview

We're doing a peculiar thing where we're relying on function equality to
decide what the button label is.

paulopmt1 commented:
### Theme `Bricksy Pro` (without custom styles)

Since we want to keep the back button at the top in this case, everything looks good!
The main change is that once we merge the StepContainerV2 changes, we'll have the back button in the top, instead of being in the bottom as it is currently on prod:

| Prod | wpcalypso | PR with flag | PR without flag |
|--------|--------|--------|--------|
| ![image](https://github.com/user-attachments/assets/3f46856e-0c4d-4404-911b-b40aa61d26a0) | ![image](https://github.com/user-attachments/assets/96cca7c6-2d4d-496e-8ba9-dc457de74274) | ![image](https://github.com/user-attachments/assets/fbc7b655-7b2b-4f75-9657-6b4cae07b33e) | ![image](https://github.com/user-attachments/assets/edb196ea-29fe-42db-9e85-cd5501f80a30) |


### Theme `Dropp` (with custom colors and fonts)

Tested in this order: | Prod | wpcalypso | PR with flag | PR without flag |
I found a minor issue with the Back CTA on minute 2:12 that only shows when we're selecting the font style or theme colors. Also happening on prod:

![image](https://github.com/user-attachments/assets/fef65b18-a3d9-4a89-b4b9-9ee933626c4c)

Otherwise, everything is looking good!

https://github.com/user-attachments/assets/01db9a9d-8c7e-40a2-a450-1050b4e5cdd4

### Potential bug with the scroll

I did try to simulate it on my real phone, but it didn't happen, so it's something from the Chrome simulator:

https://github.com/user-attachments/assets/af82012e-a616-40f1-a93c-926517c9df73

paulopmt1 commented:
I ran E2E pre-release tests in this branch, just in case, and all of them passed as well :)

p-jackson commented:
> Is the feature flag on when running these tests?

I see the feature flag enabled when I try `calypso.live`, so yes I believe the feature flag is on here. Plus the feature flag is enabled on `wpcalypso` too, where the canary tests run.

So really the issue is that I don't think there's anywhere the tests run with the feature flag off 😢

p-jackson commented:
I discovered a regression in the flow where
- flag is disabled
- initially choose free
- choose Chrono
- unlock theme

After the checkout it should transfer to atomic. But it's not.

zaguiini committed:
Fix design preview layouting for themes with only a screenshot

paulopmt1 commented:
> * Chrono

I couldn't reproduce it either, and in my case, it transferred as expected. I found another issue, though.
Slack thread here: p1743711933459729-slack-C04H4NY6STW

p-jackson committed:
Temporarily disable feature flag to run e2e tests on Calypso Live

p-jackson committed:
Reinstate feature flag

p-jackson reviewed (approved):
💪 I think we've worked out all the bugs. I didn't end up doing all 48 permutations exactly, but I believe all the flows are covered now. I'll be merging once the tests pass.
